src/Library/Lod/LodReader.cpp

```diff
@@ -25,201 +23,196 @@ static inline size_t _getDirectoryHeaderImgSize(LodVersion lod_version) {
-    auto lod = std::make_unique<LodReader>();
+    std::unique_ptr<LodReader> lod = std::make_unique<LodReader>();
+
     if (nullptr == lod) {
         Warn("LodReader::open: out of memory loading: %s", filename.c_str());
         return nullptr;
     }
 
-    auto fp = fopen(filename.c_str(), "rb");
-    if (nullptr == fp) {
+    lod->_lod = Blob::fromFile(filename);
+    if (lod->_lod.size() == 0) {
         Warn("LodReader::open: file not found: %s", filename.c_str());
         return nullptr;
     }
 
-    size_t num_expected_directories = 0;
-    bool is_lod = _lodParseHeader(fp, lod->_version, lod->_description, num_expected_directories);
-    if (!is_lod) {
+    if (lod->_lod.size() < sizeof(LodHeader_Mm6)) {
+        Warn("LodReader::open: invalid LOD file: %s", filename.c_str());
+        return nullptr;
+    }
+
+    MemoryInputStream stream(lod->_lod.data(), lod->_lod.size());
+    LodHeader_Mm6 header;
+    deserialize(stream, &header);
+
+    if (memcmp(header.signature.data(), "LOD\0", sizeof(header.signature))) {
         Warn("LodReader::open: invalid LOD file: %s", filename.c_str());
-        fclose(fp);
         return nullptr;
     }
 
-    bool is_index_ok = _lodParseDirectories(fp, lod->_version, num_expected_directories, lod->_index);
-    if (is_index_ok) {
-        lod->_fp = fp;
-        return lod;
+    if (!lodHeaderParseVersion(header, lod->_version)) {
+        Warn("LodReader::open: unknown LOD version: %s", (const char *)header.version.data());
+        return nullptr;
     }
 
-    Warn("LodReader::open: corrupt directory index: %s", filename.c_str());
-    fclose(fp);
-    return nullptr;
-}
+    if (!lod->parseDirectories(header.numDirectories, sizeof(LodHeader_Mm6))) {
+        Warn("LodReader::open: corrupt directory index: %s", filename.c_str());
+        return nullptr;
+    }
 
+    lod->_description = std::string((const char *)header.description.data());
+    return lod;
+}
 
 bool LodReader::exists(const std::string &filename) const {
-    const auto &dir = _index.front(); // only first dir is ever used, no matter names
-    return dir.files.cend() != std::find_if(
-        dir.files.cbegin(),
-        dir.files.cend(),
-        [&](const LodFile &file) {return iequals(file.name, filename); }
-    );
+    return _files.cend() != std::find_if(_files.cbegin(), _files.cend(), [&](const FileEntryDesc &file) { return iequals(file.name, filename); });
 }
 
-
 Blob LodReader::read(const std::string &filename) {
-    const auto &dir = _index.front(); // only first dir is ever used, no matter names
-    const auto &file = std::find_if(
-        dir.files.cbegin(),
-        dir.files.cend(),
-        [&](const LodFile &file) {return iequals(file.name, filename); }
-    );
-
-    if (dir.files.cend() == file) {
+    const auto &file = std::find_if(_files.cbegin(), _files.cend(), [&](const FileEntryDesc &file) { return iequals(file.name, filename); });
```

Comment:
Tbh the linear search here always irked me. Let's add a TODO to use std::unordered_map.